marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+      case find (eq rootBlockName <<< view (_1 <<< _id)) blocks of
+        Nothing -> throwError $ RootElementNotFound rootBlockName
+        Just (Tuple block nexts) ->
+          if length nexts /= 0 then
+            throwError $ IncorrectSiblingNesting "root"
+          else
+            pure block
+  where
+  -- Tries to read an Element as a Block node. It returns a tuple with the parsed Block and an Array of Block's which
+  -- are the parsed <next> elements which represent the siblings of a parent Statement.
+  readAsBlock :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m (Tuple Block (Array Block))
+  readAsBlock element =
+    if Element.tagName element /= "block" then
+      throwError $ TypeMismatch element "block"
+    else do
+      blockId <- liftEffect $ Element.id element
+      blockType <- getAttribute "type" element
+      elementChildren <- getChildrens element
+      -- Parse each child element, which can be Either a BlockChild (field, value, statement) or a list of
+      -- siblings to pass to my parent
+      { left, right } <- separate <$> traverse readAsBlockChild elementChildren
+      let
+        children = Object.fromFoldable left
+      pure
+        $ Tuple
+            (Block { id: blockId, type: blockType, children })
+            (join right)
+
+  -- Tries to read an Element as the children of a Block node. The Left hand side of the return represents a Block direct children
+  -- while the Right hand side represents the parsed `<next>` elements which should be appended to a parents Statement.
+  readAsBlockChild ::
+    forall m.
+    MonadEffect m =>
+    MonadThrow ReadDomError m =>
+    Element ->
+    m (Either ChildWithName (Array Block))
+  readAsBlockChild element = do
+    case Element.tagName element of
+      "field" -> do
+        name <- getAttribute "name" element
+        value <- getElementText element
+        pure $ Left $ Tuple name $ Field value
+      "statement" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Left $ Tuple name $ Statement ([ block ] <> nexts)
+      "value" -> do
+        name <- getAttribute "name" element
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        -- Even if a value can have a Block appended, it should have a single block child and no "siblings"
+        if length nexts /= 0 then
+          throwError $ IncorrectSiblingNesting "value"
+        else
+          pure $ Left $ Tuple name $ Value block
+      "next" -> do
+        child <- getSingleChild element
+        Tuple block nexts <- readAsBlock child
+        pure $ Right ([ block ] <> nexts)
+      _ -> throwError $ TypeMismatch element "field, statement, value, next"
+
+  getSingleChild :: forall m. MonadEffect m => MonadThrow ReadDomError m => Element -> m Element
+  getSingleChild element = do
+    children <- getChildrens element
```

Comment:
😱  Use https://pursuit.purescript.org/packages/purescript-arrays/5.3.1/docs/Data.Array#v:uncons instead. Note that the current unsafe implementation would cause a runtime exception if there were no children.

Reply:
But I check the length of the array in the step before... If it's not exactly 1 I return an expected error `SingleChildExpected`

Reply:
there's still no reason to do this in an unsafe way
```purescript
case uncons children of
  Just { head, tail: [] } -> pure head
  _ -> throwError $ SingleChildExpected element $ length children
```

Reply:
done